File: android/app/src/main/kotlin/com/yoink/net/Peer.kt
package com.yoink.net

import android.os.SystemClock
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import org.json.JSONObject

/**
 * One paired WebSocket connection, with its own heartbeat lifecycle.
 *
 * Port of daemon/net/peer.py — same intervals, so both ends agree on when a
 * link is dead. A Peer only exists once pairing has passed, so anything holding
 * one is already trusted.
 *
 * Heartbeats: send a `heartbeat` envelope every [HEARTBEAT_MS]; any inbound
 * message refreshes "last heard". Silent for [DEAD_MS] (~3 missed beats) and
 * the link is declared dead and closed. A peer whose process is killed closes
 * the socket cleanly and is noticed immediately — the heartbeat is for the
 * ungraceful case (frozen peer, dropped WiFi) where the socket is left
 * half-open. On a phone that is the common case, not the rare one.
 */
class Peer(
    val link: Link,
    val deviceId: String,
    val name: String,
    private val meName: String,
    private val onReceive: (JSONObject) -> Unit,
    private val onStatus: (String) -> Unit,
) {
    private var lastHeard = SystemClock.elapsedRealtime()

    fun send(env: JSONObject) = link.send(env)

    /** Serve this connection until it closes. Returns when the peer is gone. */
    suspend fun run() = coroutineScope {
        val heartbeat = launch {
            while (isActive) {
                delay(HEARTBEAT_MS)
                if (SystemClock.elapsedRealtime() - lastHeard > DEAD_MS) {
                    onStatus("heartbeat lost — dropping $name")
                    link.close()
                    return@launch
                }
                link.send(Protocol.envelope("heartbeat", sender = meName))
            }
        }

        try {
            while (true) {
                val env = link.next()
                lastHeard = SystemClock.elapsedRealtime()
                if (env.stringOrNull("kind") == "payload") onReceive(env)
                // heartbeats need no handling; refreshing lastHeard is the point
            }
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            // channel closed: normal end of life for this link
        } finally {
            heartbeat.cancel()
            link.close()
        }
    }

    companion object {
        const val HEARTBEAT_MS = 3_000L
        const val DEAD_MS = 10_000L
    }
}

File: mobile/lib/net/peer.dart
/// One paired WebSocket connection, with its own heartbeat lifecycle.
///
/// Port of daemon/net/peer.py — same intervals, so both ends agree on when a
/// link is dead. A Peer only exists once pairing has passed, so anything holding
/// a Peer is already trusted.
///
/// Heartbeats: send a `heartbeat` envelope every [heartbeatInterval]; any
/// inbound message refreshes "last heard". Silent for [deadAfter] (~3 missed
/// beats) and the link is declared dead and closed. A peer whose process is
/// killed closes the socket cleanly and is noticed immediately — the heartbeat
/// is for the ungraceful case (frozen peer, dropped WiFi) where the socket is
/// left half-open. On a phone that is the common case, not the rare one.
library;

import 'dart:async';

import 'link.dart';
import 'protocol.dart' as protocol;

const Duration heartbeatInterval = Duration(seconds: 3);
const Duration deadAfter = Duration(seconds: 10);

class Peer {
  Peer(
    this.link, {
    required this.deviceId,
    required this.name,
    required this.meName,
    required this.onReceive,
    required this.onStatus,
  });

  final Link link;
  final String deviceId;
  final String name;
  final String meName;
  final void Function(Map<String, dynamic> env) onReceive;
  final void Function(String msg) onStatus;

  final _clock = Stopwatch()..start();
  Duration _lastHeard = Duration.zero;

  void send(Map<String, dynamic> env) => link.send(env);

  /// Serve this connection until it closes. Returns when the peer is gone.
  Future<void> run() async {
    final hb = Timer.periodic(heartbeatInterval, (t) {
      if (_clock.elapsed - _lastHeard > deadAfter) {
        onStatus('heartbeat lost — dropping $name');
        t.cancel();
        link.close();
        return;
      }
      link.send(protocol.envelope('heartbeat', sender: meName));
    });

    try {
      while (true) {
        final env = await link.next();
        _lastHeard = _clock.elapsed;
        if (env['kind'] == 'payload') onReceive(env);
        // heartbeats need no handling; refreshing _lastHeard above is the point
      }
    } on SocketDropped {
      // normal end of life for this link
    } finally {
      hb.cancel();
      await link.close();
    }
  }
}
